gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,429 @@
+        elif device.networkAccessIdentifier:
+            key = f"{_prefix_device}:{device.networkAccessIdentifier}"
+            id = await self._get_device(key)
+        elif device.ipv4Address and device.ipv4Address.privateAddress:
+            key = f"{_prefix_device}:{device.ipv4Address.privateAddress}"
+            id = await self._get_device(key)
+        elif device.ipv4Address and device.ipv4Address.publicPort:
+            key = f"{_prefix_device}:{device.ipv4Address.publicAddress}:{device.ipv4Address.publicPort}"
+            id = await self._get_device(key)
+        elif device.ipv6Address:
+            key = f"{_prefix_device}:{device.ipv6Address.exploded}"
+            id = await self._get_device(key)
+
+        if len(id) == 0 or id is None:
+            raise ResourceNotFound()
+
+        sub_info = []
+
+        for sub_id in id:
+            sub_info.append(await self.get_qod_information_by_id(sub_id))
+
+        return sub_info
+
+    async def extend_session(self, id: str, req: ExtendSessionDuration) -> SessionInfo:
+        sub_info = await self.get_qod_information_by_id(id)
+
+        if sub_info.device is None:
+            raise ResourceNotFound()
+
+        key = f"{_prefix_gateway_nef}:{id}"
+        nef_id = await self.redis.get(key)
+
+        payload = AsSessionWithQoSSubscription(
+            self=nef_id,
+            supportedFeatures="800820",
+            notificationDestination=Link(
+                f"{self.notification_url}/callbacks/v1/qod/{id}"
+            ),
+            flowInfo=self._get_flow_description(sub_info),
+            qosReference=sub_info.qosProfile,
+            usageThreshold=UsageThreshold(
+                duration=sub_info.duration + req.requestedAdditionalDuration,
+                totalVolume=0,
+                downlinkVolume=0,
+                uplinkVolume=0,
+            ),
+        )
+
+        if sub_info.device.ipv4Address and sub_info.device.ipv4Address.publicAddress:
+            payload.ueIpv4Addr = sub_info.device.ipv4Address.publicAddress
+        elif sub_info.device.ipv6Address:
+            payload.ueIpv6Addr = sub_info.device.ipv6Address
+        elif sub_info.device.phoneNumber:
+            phone_number = sub_info.device.phoneNumber.lstrip("+")
+            payload.gpsi = f"msisdn-{phone_number}"
+
+        res = await self.httpx_client.patch(
+            nef_id,
+            content=payload.model_dump_json(exclude_unset=True),
+            headers={"Content-Type": "application/json"},
+        )
+
+        if res.status_code == 404:
+            raise ResourceNotFound()
+
+        if not res.is_success:
+            raise BadRequest()
+
+        if sub_info.expiresAt is None:
+            raise ResourceNotFound()
+
+        sub_info.expiresAt = sub_info.expiresAt + datetime.timedelta(
+            seconds=req.requestedAdditionalDuration
+        )
+
+        sub_info.duration = sub_info.duration + req.requestedAdditionalDuration
```

Comment:
You are calculating this twice, first update the values in the CAMARA subscription, and then use that value when making the NEF patch.